File: python/rafter_cli/core/surface/serialize.py
from __future__ import annotations

import json
import unicodedata
from collections.abc import Mapping, Sequence
from typing import Any

from .model import KindSpec, Transition, Unanalyzed

MAX_SAFE_INTEGER = 2**53 - 1


def _normalize_string(value: str) -> str:
    repaired: list[str] = []
    index = 0
    while index < len(value):
        codepoint = ord(value[index])
        if 0xD800 <= codepoint <= 0xDBFF:
            if index + 1 < len(value) and 0xDC00 <= ord(value[index + 1]) <= 0xDFFF:
                high = codepoint - 0xD800
                low = ord(value[index + 1]) - 0xDC00
                repaired.append(chr(0x10000 + (high << 10) + low))
                index += 2
                continue
            repaired.append("\ufffd")
        elif 0xDC00 <= codepoint <= 0xDFFF:
            repaired.append("\ufffd")
        else:
            repaired.append(value[index])
        index += 1
    return unicodedata.normalize("NFC", "".join(repaired))


def _canonicalize(value: Any) -> Any:
    if value is None or isinstance(value, bool):
        return value
    if isinstance(value, str):
        return _normalize_string(value)
    if isinstance(value, int):
        if abs(value) > MAX_SAFE_INTEGER:
            raise TypeError("canonical JSON accepts only safe integers")
        return value
    if isinstance(value, float):
        raise TypeError("canonical JSON accepts no floats")
    if isinstance(value, Mapping):
        output: dict[str, Any] = {}
        for key in sorted(value):
            if not isinstance(key, str):
                raise TypeError("canonical JSON object keys must be strings")
            if not key or any(ord(character) > 0x7F for character in key):
                raise TypeError(f"non-ASCII object key: {key}")
            output[key] = _canonicalize(value[key])
        return output
    if isinstance(value, Sequence):
        return [_canonicalize(item) for item in value]
    raise TypeError(f"unsupported JSON value: {type(value).__name__}")


def canonical_json(value: Any) -> str:
    return json.dumps(
        _canonicalize(value),
        sort_keys=True,
        ensure_ascii=False,
        separators=(",", ":"),
    )


def sort_unanalyzed(items: Sequence[Unanalyzed]) -> list[Unanalyzed]:
    return sorted(items, key=lambda item: (item.file.encode("utf-8"), item.side))


def transition_to_wire(transition: Transition) -> dict[str, Any]:
    return {
        "kind": transition.kind,
        "key": transition.key,
        "subject": transition.subject,
        "change": transition.change,
        "danger": transition.danger,
        "severity": transition.severity,
        "axes": [
            {
                "axis": axis.axis,
                "from": axis.from_rank,
                "to": axis.to_rank,
                "order": axis.order,
            }
            for axis in transition.axes
        ],
        "from": transition.from_level,
        "to": transition.to_level,
        "confidence": transition.confidence,
        "base_evidence": (
            None
            if transition.base_evidence is None
            else {"file": transition.base_evidence.file, "line": transition.base_evidence.line}
        ),
        "head_evidence": (
            None
            if transition.head_evidence is None
            else {"file": transition.head_evidence.file, "line": transition.head_evidence.line}
        ),
        "attrs": dict(transition.attrs),
        "paired": transition.paired,
    }


def kind_spec_to_wire(spec: KindSpec) -> dict[str, Any]:
    output: dict[str, Any] = {
        "kind": spec.kind,
        "comparator": spec.comparator,
        "axes": [
            {
                "name": axis.name,
                "ranks": list(axis.ranks),
                "absentRank": axis.absent_rank,
                "severityAtTop": axis.severity_at_top,
            }
            for axis in spec.axes
        ],
        "invertDanger": spec.invert_danger,
        "allowResidualPairing": spec.allow_residual_pairing,
        "display": spec.display,
    }
    if spec.severity_by_level is not None:
        output["severityByLevel"] = list(spec.severity_by_level)
    if spec.severity_when_absent is not None:
        output["severityWhenAbsent"] = spec.severity_when_absent
    if spec.severity_when_incomparable is not None:
        output["severityWhenIncomparable"] = spec.severity_when_incomparable
    return output


def kind_specs_to_wire(specs: Sequence[KindSpec]) -> list[dict[str, Any]]:
    return [kind_spec_to_wire(spec) for spec in specs]

File: node/src/core/surface/serialize.ts
import type { KindSpec, Transition, Unanalyzed } from "./model.js";

const MAX_SAFE_INTEGER = 2 ** 53 - 1;

export type JsonValue =
  | null
  | boolean
  | number
  | string
  | JsonValue[]
  | { [key: string]: JsonValue };

function normalizeString(value: string): string {
  let repaired = "";
  for (let index = 0; index < value.length; index += 1) {
    const unit = value.charCodeAt(index);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      const next = value.charCodeAt(index + 1);
      if (next >= 0xdc00 && next <= 0xdfff) {
        repaired += value[index] + value[index + 1];
        index += 1;
      } else {
        repaired += "\ufffd";
      }
    } else if (unit >= 0xdc00 && unit <= 0xdfff) {
      repaired += "\ufffd";
    } else {
      repaired += value[index];
    }
  }
  return repaired.normalize("NFC");
}

function canonicalize(value: unknown): JsonValue {
  if (value === null || typeof value === "boolean") return value;
  if (typeof value === "string") return normalizeString(value);
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value) || Math.abs(value) > MAX_SAFE_INTEGER) {
      throw new TypeError("canonical JSON accepts only safe integers");
    }
    return value;
  }
  if (Array.isArray(value)) return value.map(canonicalize);
  if (typeof value !== "object") throw new TypeError(`unsupported JSON value: ${typeof value}`);

  const input = value as Record<string, unknown>;
  const output = Object.create(null) as Record<string, JsonValue>;
  for (const key of Object.keys(input).sort()) {
    if (!/^[\x00-\x7f]+$/.test(key)) throw new TypeError(`non-ASCII object key: ${key}`);
    output[key] = canonicalize(input[key]);
  }
  return output;
}

export function canonicalJson(value: unknown): string {
  return JSON.stringify(canonicalize(value));
}

export function sortUnanalyzed(items: readonly Unanalyzed[]): Unanalyzed[] {
  return [...items].sort((left, right) => {
    const fileOrder = Buffer.compare(Buffer.from(left.file, "utf8"), Buffer.from(right.file, "utf8"));
    if (fileOrder !== 0) return fileOrder;
    return left.side === right.side ? 0 : left.side < right.side ? -1 : 1;
  });
}

export function transitionToWire(transition: Transition): JsonValue {
  return {
    kind: transition.kind,
    key: transition.key,
    subject: transition.subject,
    label: transition.label,
    change: transition.change,
    danger: transition.danger,
    severity: transition.severity,
    axes: transition.axes.map((axis) => ({
      axis: axis.axis,
      from: axis.from,
      to: axis.to,
      order: axis.order,
    })),
    from: transition.from,
    to: transition.to,
    confidence: transition.confidence,
    base_evidence: transition.baseEvidence === null
      ? null
      : { file: transition.baseEvidence.file, line: transition.baseEvidence.line },
    head_evidence: transition.headEvidence === null
      ? null
      : { file: transition.headEvidence.file, line: transition.headEvidence.line },
    attrs: { ...transition.attrs },
    paired: transition.paired,
  };
}

export function kindSpecToWire(spec: KindSpec): JsonValue {
  const output: Record<string, JsonValue> = {
    kind: spec.kind,
    comparator: spec.comparator,
    axes: spec.axes.map((axis) => ({
      name: axis.name,
      derivedFrom: [...axis.derivedFrom],
      ranks: [...axis.ranks],
      absentRank: axis.absentRank,
      severityAtTop: axis.severityAtTop,
    })),
    invertDanger: spec.invertDanger,
    allowResidualPairing: spec.allowResidualPairing,
    keyMayRepeat: spec.keyMayRepeat,
    display: spec.display,
  };
  if (spec.severityByLevel !== undefined) output.severityByLevel = [...spec.severityByLevel];
  if (spec.severityWhenAbsent !== undefined) output.severityWhenAbsent = spec.severityWhenAbsent;
  if (spec.severityWhenIncomparable !== undefined) {
    output.severityWhenIncomparable = spec.severityWhenIncomparable;
  }
  return output;
}

export function kindSpecsToWire(specs: readonly KindSpec[]): JsonValue {
  return specs.map(kindSpecToWire);
}
